src/posix/platform/infra_if.cpp

```diff
@@ -515,6 +524,81 @@ void InfraNetif::ReceiveIcmp6Message(void)
     }
 }
 
+void InfraNetif::HandleHostAddressResponse(union sigval sv)
+{
+    otError          error = OT_ERROR_NONE;
+    struct gaicb *   req   = (struct gaicb *)sv.sival_ptr;
+    struct addrinfo *res   = (struct addrinfo *)req->ar_result;
+
+    const uint8_t kMaxHostAddresses = 10;
+    otIp6Address  addresses[kMaxHostAddresses];
+    uint8_t       numAddresses = 0;
+
+    otLogInfoPlat("Handling host address response...");
+
+    for (struct addrinfo *rp = res; rp != NULL && numAddresses < kMaxHostAddresses; rp = rp->ai_next)
+    {
+        struct sockaddr_in6 *ip6Addr;
+
+        if (rp->ai_family != AF_INET6)
+        {
+            continue;
+        }
+
+        ip6Addr = reinterpret_cast<sockaddr_in6 *>(rp->ai_addr);
+        memset(&addresses[numAddresses], 0, sizeof(otIp6Address));
+        memcpy(&addresses[numAddresses].mFields.m8, &ip6Addr->sin6_addr.s6_addr, OT_IP6_ADDRESS_SIZE);
+        numAddresses++;
+    }
+
+    error = otPlatInfraIfReceiveHostAddress(gInstance, mInfraIfIndex, (char *)req->ar_name, addresses, numAddresses);
+    if (error != OT_ERROR_NONE)
+    {
+        otLogDebgPlat("Failed to handle received host address: %s", otThreadErrorToString(error));
+    }
+
+    freeaddrinfo(res);
+}
+
+otError InfraNetif::RequestHostAddressAsync(uint32_t aInfraIfIndex, const char *aHostName)
+{
+    otError          error = OT_ERROR_NONE;
+    struct addrinfo *hints;
+    struct gaicb *   reqs[1];
+    struct sigevent  sig;
+    int              status;
+
+    VerifyOrExit(aInfraIfIndex == mInfraIfIndex, error = OT_ERROR_DROP);
+
+    hints = (struct addrinfo *)malloc(sizeof(struct addrinfo));
+    memset(hints, 0, sizeof(struct addrinfo));
+    hints->ai_family   = AF_INET6;
+    hints->ai_socktype = SOCK_STREAM;
+
+    reqs[0] = (struct gaicb *)malloc(sizeof(struct gaicb));
+    memset(reqs[0], 0, sizeof(struct gaicb));
+    reqs[0]->ar_name    = aHostName;
+    reqs[0]->ar_request = hints;
+
+    memset(&sig, 0, sizeof(struct sigevent));
+    sig.sigev_notify          = SIGEV_THREAD;
+    sig.sigev_value.sival_ptr = reqs[0];
+    sig.sigev_notify_function = &InfraNetif::HandleHostAddressResponse;
+
+    status = getaddrinfo_a(GAI_NOWAIT, reqs, 1, &sig);
+
+    if (status != 0)
+    {
+        otLogNotePlat("getaddrinfo_a failed: %s", gai_strerror(status));
+        freeaddrinfo(hints);
+        ExitNow(error = OT_ERROR_FAILED);
```

Comment:
should we free hints and sig ?

Reply:
hints is freed by `freeaddinfo`.
sig is not a pointer?